File: web/src/services/stackService.ts
import api from './api';

/**
 * The stack: everything the user is carrying, dealt back one card at a time.
 * See PROBLEM.md for why the app works this way.
 */

export type StackItemStatus = 'open' | 'done' | 'dropped' | 'split';

/**
 * The two lanes: things you have to do, and things you want to do.
 * Picked once per dump session, never a gate on getting a card.
 */
export type StackKind = 'need' | 'want';

export type RankMove = 'up' | 'down' | 'top';

export const LANES: { kind: StackKind; label: string; blurb: string }[] = [
  { kind: 'need', label: 'Need to', blurb: 'The stuff that has to get done.' },
  { kind: 'want', label: 'Want to', blurb: "The stuff you'd actually like to do." },
];

export interface StackItem {
  id: string;
  title: string;
  kind: StackKind;
  status: StackItemStatus;
  position: number;
  snoozedUntil?: string | null;
  pushCount: number;
  parentId?: string | null;
  completedAt?: string | null;
  createdAt: string;
  updatedAt: string;
}

export interface StackCounts {
  /** Cards that can be dealt right now, in the requested lane. */
  remaining: number;
  /** Cards sleeping off a "Not today". */
  sleeping: number;
  /** Cards cleared, all time. */
  done: number;
  /**
   * Askable cards per lane. Lets an empty lane point at the other one without
   * ever putting a number on the card screen.
   */
  lanes: Record<StackKind, number>;
}

export interface NextCard extends StackCounts {
  /** Exactly one card, or null when the lane is clear. Never a list. */
  item: StackItem | null;
  /** The lane this card came from. */
  kind: StackKind;
  /** The card keeps getting pushed -- offer to break it down. */
  suggestSplit: boolean;
}

export interface StackCapabilities {
  /** Whether this deployment has an Anthropic key configured. */
  splitAssist: boolean;
}

export const stackService = {
  /** What this deployment can do, so the UI can hide what isn't available. */
  async getCapabilities(): Promise<StackCapabilities> {
    const response = await api.get<StackCapabilities>('/stack/capabilities');
    return response.data;
  },

  /** Job 1: get it out of the head. One thing per line, into one lane. */
  async dump(text: string, kind: StackKind = 'need'): Promise<StackCounts & { added: number }> {
    const response = await api.post('/stack/dump', { text, kind });
    return response.data;
  },

  async addOne(title: string, kind: StackKind = 'need'): Promise<StackItem> {
    const response = await api.post<StackItem>('/stack', { title, kind });
    return response.data;
  },

  /** Job 2: the whole product. One card, from one lane. */
  async getNext(kind: StackKind = 'need'): Promise<NextCard> {
    const response = await api.get<NextCard>('/stack/next', { params: { kind } });
    return response.data;
  },

  /** The full list. Always available, never the default view. */
  async getStack(
    kind: StackKind | 'all' = 'need',
    status: StackItemStatus | 'all' = 'open'
  ): Promise<StackCounts & { items: StackItem[] }> {
    const response = await api.get('/stack', { params: { kind, status } });
    return response.data;
  },

  async done(id: string): Promise<StackItem> {
    const response = await api.post<StackItem>(`/stack/${id}/done`);
    return response.data;
  },

  /** "Not now" — straight to the back, no penalty. */
  async push(id: string): Promise<StackItem> {
    const response = await api.post<StackItem>(`/stack/${id}/push`);
    return response.data;
  },

  /** "Not today" — sleeps until tomorrow, keeps its place. */
  async later(id: string): Promise<StackItem> {
    const response = await api.post<StackItem>(`/stack/${id}/later`);
    return response.data;
  },

  /** "Too big" — break it down; the first piece is dealt next. */
  async split(id: string, pieces: string): Promise<{ pieces: number }> {
    const response = await api.post(`/stack/${id}/split`, { pieces });
    return response.data;
  },

  /**
   * Ask Claude for the smallest first steps of a card. Suggests only —
   * nothing is written until the user confirms through `split`.
   */
  async suggestSplit(id: string): Promise<{ pieces: string[] }> {
    const response = await api.post<{ pieces: string[] }>(`/stack/${id}/split/suggest`);
    return response.data;
  },

  /** Move a card between the two lanes. */
  async setKind(id: string, kind: StackKind): Promise<StackItem> {
    const response = await api.post<StackItem>(`/stack/${id}/kind`, { kind });
    return response.data;
  },

  /** Order a card against its lane-mates. Always optional. */
  async rank(id: string, move: RankMove): Promise<StackItem> {
    const response = await api.post<StackItem>(`/stack/${id}/rank`, { move });
    return response.data;
  },

  async drop(id: string): Promise<StackItem> {
    const response = await api.post<StackItem>(`/stack/${id}/drop`);
    return response.data;
  },
};

File: ios/StackedWins/StackedWins/Services/StackService.swift
import Foundation

/// The stack, as the screens see it. A thin front for whichever backend is
/// active: this phone by default, a server once the user signs in to one.
///
/// Screens call these and never look behind them, so switching where the
/// stack lives is one assignment in AppState, not a change in every view.
@MainActor
enum StackService {
    /// Swapped by AppState. Starts on-device so the app works from the first
    /// open, with no account and no network.
    static var backend: StackBackend = LocalStackStore.shared

    // MARK: Job 1 — get it out of the head

    static func dump(_ text: String, into kind: StackKind) async throws -> DumpResult {
        try await backend.dump(text, into: kind)
    }

    // MARK: Job 2 — one card

    static func next(in kind: StackKind) async throws -> NextCard {
        try await backend.next(in: kind)
    }

    static func capabilities() async throws -> StackCapabilities {
        try await backend.capabilities()
    }

    /// The full list. Always reachable, never the default view.
    static func everything(in kind: StackKind) async throws -> StackList {
        try await backend.everything(in: kind)
    }

    // MARK: The four moves

    @discardableResult
    static func done(_ id: String) async throws -> StackItem {
        try await backend.done(id)
    }

    /// "Not now" — to the back of its lane, no penalty.
    @discardableResult
    static func push(_ id: String) async throws -> StackItem {
        try await backend.push(id)
    }

    /// "Not today" — sleeps until tomorrow, keeps its place.
    @discardableResult
    static func later(_ id: String) async throws -> StackItem {
        try await backend.later(id)
    }

    /// "Too big" — break it into pieces; the first piece is dealt next.
    static func split(_ id: String, pieces: String) async throws -> SplitResult {
        try await backend.split(id, pieces: pieces)
    }

    /// Ask Claude for the smallest first steps. Suggests only — nothing is
    /// written until the user confirms through `split`. Server only.
    static func suggestSplit(_ id: String) async throws -> SplitSuggestion {
        try await backend.suggestSplit(id)
    }

    // MARK: Lanes, ranking, letting go

    @discardableResult
    static func move(_ id: String, to kind: StackKind) async throws -> StackItem {
        try await backend.move(id, to: kind)
    }

    @discardableResult
    static func rank(_ id: String, _ move: RankMove) async throws -> StackItem {
        try await backend.rank(id, move)
    }

    @discardableResult
    static func drop(_ id: String) async throws -> StackItem {
        try await backend.drop(id)
    }
}
